SIPS/sip-31.md

```diff
@@ -0,0 +1,77 @@
+---
+sip: 31
+title: Client-only RPC Entrypoint
+status: Draft
+author: Daniel Rocha (@danroc)
+created: 2025-03-19
+---
+
+## Abstract
+
+This proposal introduces a new entrypoint, `onClientRequest`, allowing Snaps to
+expose a handler that can only be called by clients.
+
+While the request `origin` can currently be used to differentiate requests from
+clients and dapps, this new entrypoint increases separation and minimizes
+the risks associated with `origin` spoofing or request routing bugs.
+
+## Motivation
+
+Currently, Snaps rely on the `origin` field of incoming requests to
+differentiate whether a request originates from the client or an external dapp.
+However, this approach has potential risks:
+
+- **Origin Spoofing**: Bugs or vulnerabilities in request validation could
+  allow dapps to masquerade as clients.
+
+- **Unintended Exposure**: If a Snap processes requests without strict
+  validation, it may unintentionally expose to dapps methods that are meant
+  only for clients.
+
+- **Cleaner Separation**: Having a dedicated entrypoint for client requests
+  enforces stricter request isolation at the API level.
+
+By introducing `onClientRequest`, we ensure that Snaps can define handlers
+exclusively accessible by clients, reducing attack vectors.
+
+## Specification
+
+This proposal introduces a new optional handler function, `onClientRequest`,
+which a Snap MAY implement.
+
+### Language
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
+"SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" written
+in uppercase in this document are to be interpreted as described in [RFC
+2119](https://www.ietf.org/rfc/rfc2119.txt).
+
+### `onClientRequest` Entrypoint
```

Comment:
Should we also define a new endowment?

Reply:
We don't need it if this is restricted to pre-installed Snaps right?

Reply:
Yes, but should we limit this entrypoint to pre-installed Snaps?

Reply:
Well I'm only speculating so maybe this is not relevant, but my thinking is: 

1. We only plan to implement `onClientRequest` in flows for pre-installed Snaps. We don't have any plans to do this with any 3rd party Snaps yet. 
2. The biggest risk, which may or may not be relevant, is if a 3rd party Snap spoofs itself as a pre-installed Snap (some 3rd party Snap pretends to be the ENS Resolver Snap, or MMI Snap, for example), and thus gains malicious access. 

So this is why I think we either need to introduce an endowment or make it only for pre-installed Snaps. I don't think we can skip both options. But would like someone else to validate or correct my thinking here...

Reply:
Is there a use-case for `onClientRequest` beyond preinstalled Snaps? 

My assumption was that we wanted `onClientRequest` for non-standard JSON-RPC requests sent to Snaps in one way or another affiliated with MetaMask. Generally standardized JSON-RPC for everyone (third parties and preinstalled Snaps) should have be separate standardized handlers (e.g. `onProtocolRequest`).

Reply:
OK so sounds like we are on board to make this only for pre-installed Snaps. Sounds good to me.